rosette/src/BaseSupp.cc

```diff
@@ -521,6 +523,8 @@ DEF("char*->string", char_star_to_string, 1, 1) {
         return RBLstring::create((char*)addr);
     else
         PRIM_ERROR("invalid address");
+
+    return FIXNUM(0);
```

Comment:
Can you explain why you return `FIXNUM(0)` here?

Reply:
There's also return of the thing returned by `PRIM_ERROR`, which is a tagged type:
Before:
```C++
Prim* char_star_to_string;
pOb _f_char_star_to_string(Ob * __PRIM__, pCtxt __CTXT__);
static BuiltinPrimRecord _pr_char_star_to_string = { "char*->string", _f_char_star_to_string, &char_star_to_string, (1), (1), 0};
static BuiltinPrim _i_char_star_to_string( &_pr_char_star_to_string);
pOb _f_char_star_to_string(Ob * __PRIM__, pCtxt __CTXT__) {
    if (!(TAG(__CTXT__->argvec->elem(0)) == OTfixnum))
        return __PRIM__->mismatch(__CTXT__, (0), "Address");

    uint32_t addr = ((int)(Ob *)(TAGVAL(__CTXT__->argvec->elem(0))));;
    if (addr >= local_page_size)
        return RBLstring::create((char*)addr);
    else
        __PRIM__->runtimeError(__CTXT__, "invalid address");
}
```
After
```C++
Prim* char_star_to_string;
pOb _f_char_star_to_string(Ob * __PRIM__, pCtxt __CTXT__);
static BuiltinPrimRecord _pr_char_star_to_string = { "char*->string", _f_char_star_to_string, &char_star_to_string, (1), (1), 0};
static BuiltinPrim _i_char_star_to_string( &_pr_char_star_to_string);
pOb _f_char_star_to_string(Ob * __PRIM__, pCtxt __CTXT__) {
    if (!(TAG(__CTXT__->argvec->elem(0)) == OTfixnum))
        return __PRIM__->mismatch(__CTXT__, (0), "Address");

    uint32_t addr = ((int)(Ob *)(TAGVAL(__CTXT__->argvec->elem(0))));;
    if (addr >= local_page_size)
        return RBLstring::create((char*)addr);
    else
        __PRIM__->runtimeError(__CTXT__, "invalid address");

    return ((pOb)(((unsigned)(0) << TagSize) | (unsigned)OTfixnum));
}
```
Where `Ob::runtimeError()` returns `DEADTHREAD`, which is
```C++
#define DEADTHREAD MAKEESCTAGGED(OTsysval, syscodeDeadThread)
```

Reply:
Sorry, I still don't understand why you want to return `FIXNUM(0)` there. `char*->string` either returns an `RBLString` or whatever is returned by the `PRIM_ERROR` macro. What is wrong about that?

Reply:
Part of my point was that I don't think the original code returned anything in the else branch. Which probably means it returned 0 (or `pOb(0)` which looks like `NULL`) . . . the macro resolves to a function call, not a `return`.

Reply:
I guess I assumed the the `runtimeError()` thing set up some sort of rosette-level non-local exit, but I don't have the wherewithal to trace through that. Either way, `FIXNUM(0)` is likely different.

Reply:
This return can't ever execute, so instead of this:

```
    if (addr >= local_page_size)
        return RBLstring::create((char*)addr);
    else
        PRIM_ERROR("invalid address");

    return FIXNUM(0);
```

Please write this:

```
    if (addr >= local_page_size) {
        return RBLstring::create((char*)addr);
    }
    PRIM_ERROR("invalid address");
```

I am unsure as to whether it makes sense to `return PRIM_ERROR(...)` or to `return FIXNUM(0)`, but both of them are a change in behavior. My guess is that wherever a primitive is being called from checks the for a run-time error before examining the return, otherwise there would have been some fun bugs back in the day. Keith, can you confirm or do we have a zillion call sites for this?

Reply:
FIXNUM(0) was just my best guess.  It would appear that I was wrong (big surprise there).

I'll take another look at this.  Where it makes sense, I think the right thing to do is return the results of PRIM_ERROR.

Not returning anything would result in the return value being random left over data, which I'm sure is not what we want.

Reply:
@kjmcqueen @tymm I think for every instance of the "no return" warning, we should `return PRIM_ERROR(...)` if there is one. There is no consistent behavior that rosette _could be relying on_, so the only risk here is that we expose a bug that would have been harder to detect and fix later on.

Reply:
@kjmcqueen @nashef `Ob::runtimeError(...)` (which is called by `PRIM_ERROR(...)`) changes the state of the VM and there are checks on the call site whether `DEADTHREAD` is returned or not. Therefore I see a risk of introducing unwanted behavior when we do `return PRIM_ERROR(...)` everywhere where the "no return" warning is popping up.

Reply:
There is broad agreement in Discord to switch to deterministic behavior, i.e. returning something explicitly, rather than rely on undefined behavior. So, please do the `return PRIM_ERROR(...)` thing wherever the compiler warns.

Reply:
@tymm requests that you add a comment to indicate that you have done this.

Reply:
**Final:**  We will use `return PRIM_ERROR(...)` with an appropriate comment. This may result in a change in behavior, but it will be deterministic.